src/libpsl-native/src/createprocess.cpp

```diff
@@ -0,0 +1,197 @@
+#include "createprocess.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <limits>
+#include <limits.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <sys/resource.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <syslog.h>
+#include <unistd.h>
+
+enum
+{
+    SUPPRESS_PROCESS_SIGINT = 0x00000001
+};
+
+enum
+{
+    READ_END_OF_PIPE = 0,
+    WRITE_END_OF_PIPE = 1
+};
+
+static void CloseIfOpen(int fd)
+{
+    if (fd >= 0)
+    {
+        close(fd); // Ignoring errors from close is a deliberate choice
+    }
+}
+
+// Checks if the IO operation was interrupted and needs to be retried.
+// Returns true if the operation was interrupted; otherwise, false.
+template <typename TInt>
+static inline bool CheckInterrupted(TInt result)
+{
+    return result < 0 && errno == EINTR;
+}
+
+static int Dup2WithInterruptedRetry(int oldfd, int newfd)
+{
+    int result;
+    while (CheckInterrupted(result = dup2(oldfd, newfd)));
+    return result;
+}
+
+int32_t ForkAndExecProcess(
+    const char* filename,
+    char* const argv[],
+    char* const envp[],
+    const char* cwd,
+    int32_t redirectStdin,
+    int32_t redirectStdout,
+    int32_t redirectStderr,
+    int32_t creationFlags,
+    int32_t* childPid,
+    int32_t* stdinFd,
+    int32_t* stdoutFd,
+    int32_t* stderrFd)
+{
+    int success = true;
+    int processId = -1;
+    int stdinFds[2] = { -1, -1 };
+    int stdoutFds[2] = { -1, -1 };
+    int stderrFds[2] = { -1, -1 };
+
+    // Validate arguments
+    if (nullptr == filename || nullptr == argv || nullptr == envp || nullptr == stdinFd || nullptr == stdoutFd ||
+        nullptr == stderrFd || nullptr == childPid)
+    {
+        assert(false && "null argument.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    if ((redirectStdin & ~1) != 0 || (redirectStdout & ~1) != 0 || (redirectStderr & ~1) != 0)
+    {
+        assert(false && "Boolean redirect* inputs must be 0 or 1.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    // Make sure we can find and access the executable. exec will do this, of course, but at that point it's already
+    // in the child process, at which point it'll translate to the child process' exit code rather than to failing
+    // the Start itself.  There's a race condition here, in that this could change prior to exec's checks, but there's
+    // little we can do about that. There are also more rigorous checks exec does, such as validating the executable
+    // format of the target; such errors will emerge via the child process' exit code.
+    if (access(filename, X_OK) != 0)
+    {
+        success = false;
+        goto done;
+    }
+
+    // Open pipes for any requests to redirect stdin/stdout/stderr
+    if ((redirectStdin && pipe(stdinFds) != 0) || (redirectStdout && pipe(stdoutFds) != 0) ||
+        (redirectStderr && pipe(stderrFds) != 0))

```

Comment:
Take the write end of the redirected stdin pipe as an instance (similar to the redirected stdout/stderr pipes), a FD pointing to the write end of that pipe will be created in powershell process after creating the SSH process. Assuming powershell then creates some other processes before SSH process terminates (I'm thinking of the scenario where I create a SSH PSSession but no an interactive remote session, so it's possible to run native commands locally while the SSH process is still alive), then that FD will be inherited by all those processes, so they are all potentially able to write to that pipe or affect that pipe in possible ways.

In this case, the SSH process is the only reader of that pipe, so when SSH terminates, "_the only reader dies, and then the writer gets notified about this by getting a SIGPIPE or at least an EPIPE error (depending on how signals are defined)._" (quoted from [here](https://stackoverflow.com/questions/19265191/why-should-you-close-a-pipe-in-linux))

I'm not saying it will definitely cause a problem, but just that if the leaked FD cause trouble to us, it might be very hard to figure out the real cause.